Datagram speedtesting

dr7ana opened the pull request:
- implemented datagram speedtest binaries in `tests/`
- optimized send logic and dropped superfluous flags (checksum, hash, etc)
- datagrams are fast

dr7ana committed:
Datagram speedtesting

- implemented datagram speedtest binaries in `tests/`
- optimized send logic and dropped superfluous flags (checksum, hash, etc)
- datagrams are fast

jagerman commented on tests/dgram-speed-client.cpp:
```diff
@@ -0,0 +1,229 @@
+        log::critical(test_cat, "Stream {} (rawid={}) closed (error={})", i, s.stream_id(), errcode);
+    };
+
+    dgram_data_callback recv_dgram_cb = [&](bstring data) {
```
Looking at this as a sample app implementation, it seems that we have a couple issues with the argument passed here:
- when we don't have a split packet (because either we don't have packet splitting, or we receive a small packet) we are still allocating a `bstring` that we could avoid.
- there's no way to tell *which* connection sent me this datagram.  I was looking to make this server persistent, but right now I can't which is going to be a problem when this is used in a many-to-one client-to-server architecture.

The second issue is easy to fix (just pass the connection_interface along), but the first issue is trickier because *sometimes* we want to pass a `bstring` that we already have on hand (i.e. from reassembling a split packet), but other times we we don't.

dr7ana replied:
Resolving this per off-github discussion, to be addressed in #29 ongoing

jagerman committed:
Rip out hashing from dgram speed test

It isn't doing anything useful in the dgram speed test, so just remove
it.

dr7ana committed:
Exposed reply mechanism in datagram callbacks

tewinget commented on include/quic/datagram.hpp:
```diff
@@ -9,9 +9,45 @@ namespace oxen::quic
+            reply(view, std::move(keep_alive));
+        }
+
+        void reply(bstring_view data, std::shared_ptr<void> keep_alive = nullptr);
```
This should probably be renamed as it is used for a non-reply send as well as a reply.

dr7ana replied:
That's valid, but from the perspective of the callback, having "reply" somewhere makes it clear that it's directly back along that connection

From that datagram interface, the user has a getter method to retrieve the connection interface and call `send_datagram` from there

jagerman replied:
I think `reply` is fine, but should the `send_datagram` in here taking a vector of char's also be renamed to `reply`?

dr7ana replied:
That's a good point, I just used the same templating we had on connection_interface and stream for consistency. Maybe someone wants to send a big reply?

jagerman committed:
Simplify dgid calculation

We were calculating the dgid by tracking the last dgid in use, and
including the split/side bits in the last dgid, both of which complicate
the math a little.

This changes it to track the *next* dgid and doesn't track the
split/side bits, which makes it easier to understand.

This also (serendipitously) corrects a minor issue with numbering the
first packet: instead of numbering it as either 0 or 2 (depending on
whether split) we would number it as 4 or 2 (effectively skipping dgid 0
for an unsplit packet).

jagerman committed:
Fix default buf size

I don't think the default value here was actually *used* as we always
use the value from `opt::enable_datagrams`, but was misleading as it
disgrees with the actual 4096 default from `enable_datagrams`.

jagerman committed:
Memory reduction for buffers

This replaces the `received_datagram` vector with a
`std::unique_ptr<received_datagram>` vector, with a fixed buffer inside
the type for storage of the half-received datagram, rather than a
`bstring`.  Rather than have a `.clear()`, we now simply reset the
unique_ptr when we're done with it.

In typical usage, this reduces base (i.e. no actual split dgrams held)
memory usage from needing a constant 40 bytes per idle buffer entry
(160kiB per connection buffer) to 8 (32kiB), which reduces the buffer
overhead when we have active 2000 connections (i.e. SN full mesh) from
320MB to 64MB.

This also slightly reduces the data sizes of some related members to
slim down the memory usage slightly further.

(We have some further buffer rework planned to reduce this by another
factor of 4 by only needing 1024 buffer entries to maintain our same
current minimum packet reorder resilience.)

dr7ana commented on src/messages.cpp:
```diff
@@ -175,8 +175,10 @@ namespace oxen::quic
+        else if (b)
+            return {*payload, pload_id, -1, false};
         else
-            return {(b ? *payload : *addendum), (b ? pload_id : *add_id), (b ? -1 : 1), false};
```
RIP double inline ternary

jagerman commented on include/quic/connection.hpp:
```diff
@@ -17,8 +17,12 @@
 #include "types.hpp"
 #include "utils.hpp"
 
+extern std::atomic<bool> datagram_test_enabled;
```
This should be present only in a debug build

dr7ana replied:
but we dont speedtest on debug builds?

jagerman commented on include/quic/connection.hpp:
```diff
@@ -45,15 +49,19 @@ namespace oxen::quic
+    class perf_testing
+    {
+      public:
+    };
```
What is this?

jagerman commented on include/quic/opt.hpp:
```diff
@@ -48,6 +51,7 @@ namespace oxen::quic::opt
         Splitting mode = Splitting::NONE;
         // Note: this is the size of the entire buffer, divided amongst 4 rows
         int bufsize = 4096;
+        // If setr to
```
?

jagerman commented on include/quic/utils.hpp:
```diff
@@ -42,6 +42,7 @@ extern "C"
 #include <string>
 #include <string_view>
 #include <unordered_set>
+#include <variant>
```
Not used in this header, afaict?

jagerman commented on src/connection.cpp:
```diff
@@ -32,6 +32,8 @@ extern "C"
 #include "utils.hpp"
 
+std::atomic<bool> datagram_test_enabled = false;
+
```
NDEBUG

jagerman commented on tests/CMakeLists.txt:
```diff
@@ -31,6 +31,11 @@ target_link_libraries(alltests PUBLIC quic Catch2::Catch2 CLI11::CLI11)
+    add_executable(${x} ${x}.cpp)
+    target_link_libraries(${x} PRIVATE quic CLI11::CLI11)
+endforeach()
+
```
Would be simpler to just add these two executables into the loop just below

dr7ana committed:
Review fixes

jagerman committed:
Add ARM64 debug build
